When repainting after layout, compute the outline bounds in the same pass as the clipped overflow rect

smfr opened the pull request:
#### b088578e1f36d5fb9ce91269eab482bbc52be91d
<pre>
When repainting after layout, compute the outline bounds in the same pass as the clipped overflow rect
<a href="https://bugs.webkit.org/show_bug.cgi?id=265761">https://bugs.webkit.org/show_bug.cgi?id=265761</a>
<a href="https://rdar.apple.com/119101912">rdar://119101912</a>

Reviewed by Alan Baradlay.

Currently repaints originating in RenderElement::repaintAfterLayoutIfNeeded() are computing using two rectangles:
the clippedOverflowRect, and the outlineBoundsForRepaint. These are computed in two separate ancestor traversals,
but we can make things more efficient by computing them at the same time.

The core functions that compute both rects, `computeRects()` and the virtual `computeVisibleRectsInContainer()`
that it calls were updated in previous commits (271364@main, 271422@main) to handle both rectangles. These
are generic functions that compute rectangles both for repaint, and other purposes.

Here we add a new, repaint-specific wrapper, `rectsForRepaintingAfterLayout()`, which returns both rectangles
via RepaintRects. Unfortunately, a number of classes specialize the calling function, `clippedOverflowRect()`,
so we need `rectsForRepaintingAfterLayout()` overrides in those classes to replicate their special behavior.

The previous `localRectForRepaint()` becomes `localRectsForRepaint()` to return the pair of rects,
optionally including the outline bounds rect.

RenderText and RenderInline are special becuase they never cares about the outline bounds rect. RenderTableRow and
RenderTableCol are special, and SVG has some special-case behavior in the non-LBSE code path.

LayoutRepainter now stores a RepaintsRect, and knows that the outline bounds is not needed if this is a full repaint.

RenderElement::repaintAfterLayoutIfNeeded() is adjusted to do the border radius computation using the outline
bounds rect (as a proxy for the border box size). Doing this computation in the clipped rect didn&apos;t make much sense.

RenderLayer no longer needs its own RepaintRects class. Also remove the assertions about invalid repaint rects;
they fired too often, and didn&apos;t clearly reflect real bugs.

There is a minor behavior change here that affects two tests: the results of the outline bounds rect computation
are snapped to device pixels, which did not happen before. This turns some full repaints into non-full repaints,
affecting fast/repaint/hidpi-transform-on-subpixel-repaintrect.html and svg/repaint/mask-object-bounding-box-shrink.html.
Testing shows that the former still repaints correctly. The latter fails to repaint, but I think that reflects
the fact that SVG should probably treat all post-layout repaints as full repaints (webkit.org/b/265723).

* LayoutTests/fast/repaint/hidpi-transform-on-subpixel-repaintrect-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt:
* Source/WebCore/rendering/LayoutRepainter.cpp:
(WebCore::LayoutRepainter::LayoutRepainter):
(WebCore::LayoutRepainter::repaintAfterLayout):
* Source/WebCore/rendering/LayoutRepainter.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::localOutlineBoundsRepaintRect const):
(WebCore::RenderBox::outlineBoundsForRepaint const):
(WebCore::RenderBox::localRectsForRepaint const):
(WebCore::RenderBox::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderInline.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::setRepaintRects):
(WebCore::RenderLayer::recursiveUpdateLayerPositionsAfterScroll):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::repaintRects const):
* Source/WebCore/rendering/RenderLineBreak.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::localRectsForRepaint const):
(WebCore::RenderObject::rectsForRepaintingAfterLayout const):
(WebCore::RenderObject::clippedOverflowRect const):
(WebCore::RenderObject::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::RepaintRects::RepaintRects):
(WebCore::RenderObject::outlineBoundsForRepaint const):
(WebCore::RenderObject::computeRectForRepaint const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::localRectsForRepaint const):
(WebCore::RenderReplaced::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::localRectsForRepaint const):
(WebCore::RenderTableCell::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCol.cpp:
(WebCore::RenderTableCol::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderTableCol.h:
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderTableRow.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderText.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGGradientStop.h:
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::localRectsForRepaint const):
(WebCore::RenderSVGModelObject::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::localClippedOverflowRect const):
(WebCore::LegacyRenderSVGRoot::clippedOverflowRect const):
(WebCore::LegacyRenderSVGRoot::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:

Canonical link: <a href="https://commits.webkit.org/271495@main">https://commits.webkit.org/271495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62c10fe9a6fd5d151bde49dcd507e9ab4a013ca1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28563 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4578 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28833 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5975 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24586 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5206 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25587 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26175 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26032 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/32202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5299 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3468 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6904 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6840 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5760 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -907,18 +909,23 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
+    LayoutRect computeRectForRepaint(const LayoutRect& rect, const RenderLayerModelObject* repaintContainer) const { return computeRects({ rect }, repaintContainer, visibleRectContextForRepaint()).clippedOverflowRect; }
     FloatRect computeFloatRectForRepaint(const FloatRect&, const RenderLayerModelObject* repaintContainer) const;
 
+    virtual RepaintRects rectsForLayoutRepainting(const RenderLayerModelObject* repaintContainer, RepaintOutlineBounds) const;
```
what does "ForLayout" here specifically mean? (is there going to be a rectsForStyleChangeRepainting? I know what the answer is, I am just trying to figure out the reasoning for ForLayout).

smfr replied:
The "for layout" means that the consumer is `RenderElement::repaintAfterLayoutIfNeeded()`, which is the only repaint function that takes the pair of rects (compare with `computeRectForRepaint()`).

smfr replied:
I didn't use `rectsForPostLayoutRepainting()` because we compute these rects before we layout a specific renderer, but maybe that makes more sense?

smfr committed:
When repainting after layout, compute the outline bounds in the same pass as the clipped overflow rect
https://bugs.webkit.org/show_bug.cgi?id=265761
rdar://119101912

Reviewed by Alan Baradlay.

Currently repaints originating in RenderElement::repaintAfterLayoutIfNeeded() are computing using two rectangles:
the clippedOverflowRect, and the outlineBoundsForRepaint. These are computed in two separate ancestor traversals,
but we can make things more efficient by computing them at the same time.

The core functions that compute both rects, `computeRects()` and the virtual `computeVisibleRectsInContainer()`
that it calls were updated in previous commits (271364@main, 271422@main) to handle both rectangles. These
are generic functions that compute rectangles both for repaint, and other purposes.

Here we add a new, repaint-specific wrapper, `rectsForRepaintingAfterLayout()`, which returns both rectangles
via RepaintRects. Unfortunately, a number of classes specialize the calling function, `clippedOverflowRect()`,
so we need `rectsForRepaintingAfterLayout()` overrides in those classes to replicate their special behavior.

The previous `localRectForRepaint()` becomes `localRectsForRepaint()` to return the pair of rects,
optionally including the outline bounds rect.

RenderText and RenderInline are special becuase they never cares about the outline bounds rect. RenderTableRow and
RenderTableCol are special, and SVG has some special-case behavior in the non-LBSE code path.

LayoutRepainter now stores a RepaintsRect, and knows that the outline bounds is not needed if this is a full repaint.

RenderElement::repaintAfterLayoutIfNeeded() is adjusted to do the border radius computation using the outline
bounds rect (as a proxy for the border box size). Doing this computation in the clipped rect didn't make much sense.

RenderLayer no longer needs its own RepaintRects class. Also remove the assertions about invalid repaint rects;
they fired too often, and didn't clearly reflect real bugs.

There is a minor behavior change here that affects two tests: the results of the outline bounds rect computation
are snapped to device pixels, which did not happen before. This turns some full repaints into non-full repaints,
affecting fast/repaint/hidpi-transform-on-subpixel-repaintrect.html and svg/repaint/mask-object-bounding-box-shrink.html.
Testing shows that the former still repaints correctly. The latter fails to repaint, but I think that reflects
the fact that SVG should probably treat all post-layout repaints as full repaints (webkit.org/b/265723).

* LayoutTests/fast/repaint/hidpi-transform-on-subpixel-repaintrect-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt:
* Source/WebCore/rendering/LayoutRepainter.cpp:
(WebCore::LayoutRepainter::LayoutRepainter):
(WebCore::LayoutRepainter::repaintAfterLayout):
* Source/WebCore/rendering/LayoutRepainter.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::localOutlineBoundsRepaintRect const):
(WebCore::RenderBox::outlineBoundsForRepaint const):
(WebCore::RenderBox::localRectsForRepaint const):
(WebCore::RenderBox::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderInline.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::setRepaintRects):
(WebCore::RenderLayer::recursiveUpdateLayerPositionsAfterScroll):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::repaintRects const):
* Source/WebCore/rendering/RenderLineBreak.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::localRectsForRepaint const):
(WebCore::RenderObject::rectsForRepaintingAfterLayout const):
(WebCore::RenderObject::clippedOverflowRect const):
(WebCore::RenderObject::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::RepaintRects::RepaintRects):
(WebCore::RenderObject::outlineBoundsForRepaint const):
(WebCore::RenderObject::computeRectForRepaint const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::localRectsForRepaint const):
(WebCore::RenderReplaced::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::localRectsForRepaint const):
(WebCore::RenderTableCell::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCol.cpp:
(WebCore::RenderTableCol::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderTableCol.h:
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderTableRow.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/RenderText.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGGradientStop.h:
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::localRectsForRepaint const):
(WebCore::RenderSVGModelObject::localRectForRepaint const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::localClippedOverflowRect const):
(WebCore::LegacyRenderSVGRoot::clippedOverflowRect const):
(WebCore::LegacyRenderSVGRoot::rectsForRepaintingAfterLayout const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:

Canonical link: https://commits.webkit.org/271495@main
